query.go

```diff
@@ -82,174 +161,212 @@ func (dht *IpfsDHT) runDisjointQueries(ctx context.Context, d int, target string
+		completed: completed,
+	}
+
+	for i, p := range sortedPeers {
+		res.state[i] = peerState[p]
+	}
 
 	return res
 }
 
-// strictParallelismQuery concurrently sends the query RPC to all eligible peers
-// and waits for ALL the RPC's to complete before starting the next round of RPC's.
-func strictParallelismQuery(q *query) {
-	foundCloser := false
-	for {
-		// get the unqueried peers from among the K closest peers to the key sorted in ascending order
-		// of their 'distance-latency` score.
-		// We sort peers like this so that "better" peers are chosen to be in the α peers
-		// which get queried from among the unqueried K  closet.
-		peersToQuery := q.localPeers.UnqueriedFromKClosest(q.scorePeerByDistanceAndLatency,
-			func(i1 peerheap.Item, i2 peerheap.Item) bool {
-				return i1.Value.(*big.Int).Cmp(i2.Value.(*big.Int)) == -1
-			})
-
-		// The lookup terminates when the initiator has queried and gotten responses from the k
-		// closest nodes it has heard about.
-		if len(peersToQuery) == 0 {
-			return
-		}
+type queryUpdate struct {
+	seen        []peer.ID
+	queried     []peer.ID
+	unreachable []peer.ID
+}
 
-		// Of the k nodes the initiator has heard of closest to the target,
-		// it picks α that it has not yet queried and resends the FIND NODE RPC to them.
-		numQuery := q.dht.alpha
-
-		// However, If a round of RPC's fails to return a node any closer than the closest already heard about,
-		// the initiator resends the RPC'S to all of the k closest nodes it has
-		// not already queried.
-		if !foundCloser {
-			numQuery = len(peersToQuery)
-		} else if pqLen := len(peersToQuery); pqLen < numQuery {
-			// if we don't have α peers, pick whatever number we have.
-			numQuery = pqLen
-		}
+func (q *query) runWithGreedyParallelism() {
+	pathCtx, cancelPath := context.WithCancel(q.ctx)
+	defer cancelPath()
 
-		// reset foundCloser to false for the next round of RPC's
-		foundCloser = false
+	alpha := q.dht.alpha
 
-		queryResCh := make(chan *queryResult, numQuery)
-		resultsReceived := 0
+	ch := make(chan *queryUpdate, alpha)
+	ch <- &queryUpdate{seen: q.seedPeers}
 
-		// send RPC's to all the chosen peers concurrently
-		for _, p := range peersToQuery[:numQuery] {
-			go func(p peer.ID) {
-				queryResCh <- q.queryPeer(p)
-			}(p)
+	for {
+		select {
+		case update := <-ch:
+			q.updateState(update)
+		case <-pathCtx.Done():
+			q.terminate()
 		}
 
-	loop:
-		// wait for all outstanding RPC's to complete before we start the next round.
-		for {
-			select {
-			case res := <-queryResCh:
-				foundCloser = foundCloser || res.foundCloserPeer
-				resultsReceived++
-				if resultsReceived == numQuery {
-					break loop
-				}
-			case <-q.ctx.Done():
+		// termination is triggered on end-of-lookup conditions or starvation of unused peers
+		if q.readyToTerminate() {
+			q.terminate()
+
+			// exit once all goroutines have been cleaned up
+			if q.queryPeers.NumWaiting() == 0 {
 				return
 			}
+			continue
+		}
```

Comment:
It'd be nice to know when the entire query has shut down (e.g. so we can close a channel that returns the values we're finding during the query). We might want to use terminate to trigger a context cancellation that kills the other queries.

However, since we may want to finish those queries anyway as part of making sure we've queried the top k (as opposed to beta) peers this seems like a reasonable solution for now.